admin_manual/installation/nginx.rst

```diff
@@ -114,75 +76,110 @@ webroot of your nginx installation. In this example it is
+      
+      # HTTP response headers borrowed from NextCloud `.htaccess`
+      add_header Referrer-Policy                      "no-referrer"   always;
+      add_header X-Content-Type-Options               "nosniff"       always;
+      add_header X-Download-Options                   "noopen"        always;
+      add_header X-Frame-Options                      "SAMEORIGIN"    always;
+      add_header X-Permitted-Cross-Domain-Policies    "none"          always;
+      add_header X-Robots-Tag                         "none"          always;
+      add_header X-XSS-Protection                     "1; mode=block" always;
+      
+      # Remove X-Powered-By, which is an information leak
+      fastcgi_hide_header X-Powered-By;
+      
+      # Path to the root of your installation
+      root /var/www/nextcloud;
+      
+      # Specify how to handle directories -- specifying `/index.php$request_uri`
+      # here as the fallback means that Nginx always exhibits the desired behaviour
+      # when a client requests a path that corresponds to a directory that exists
+      # on the server. In particular, if that directory contains an index.php file,
+      # that file is correctly served; if it doesn't, then the request is passed to
+      # the front-end controller. This consistent behaviour means that we don't need
+      # to specify custom rules for certain paths (e.g. images and other assets,
+      # `/updater`, `/ocm-provider`, `/ocs-provider`), and thus
+      # `try_files $uri $uri/ /index.php$request_uri`
+      # always provides the desired behaviour.
+      index index.php index.html /index.php$request_uri;
+      
+      # Default Cache-Control policy
+      expires 1m;
+      
+      # Rule borrowed from `.htaccess` to handle Microsoft DAV clients
+      location = / {
+          if ( $http_user_agent ~ DavClnt ) {
+              return 302 /remote.php/webdav/$is_args$args;
+          }
       }
-
-      location ~ ^\/(?:build|tests|config|lib|3rdparty|templates|data)\/ {
-          deny all;
+      
+      location = /robots.txt {
+          allow all;
+          log_not_found off;
+          access_log off;
       }
-      location ~ ^\/(?:\.|autotest|occ|issue|indie|db_|console) {
-          deny all;
+      
+      # Make a regex exception for `/.well-known` so that clients can still
+      # access it despite the existence of the regex rule
+      # `location ~ /(\.|autotest|...)` which would otherwise handle requests
+      # for `/.well-known`.
+      location ^~ /.well-known {
+          # The following 6 rules are borrowed from `.htaccess`
+      
+          rewrite ^/\.well-known/host-meta\.json  /public.php?service=host-meta-json  last;
+          rewrite ^/\.well-known/host-meta        /public.php?service=host-meta       last;
+          rewrite ^/\.well-known/webfinger        /public.php?service=webfinger       last;
+          rewrite ^/\.well-known/nodeinfo         /public.php?service=nodeinfo        last;
```

Comment:
Not critical we figure this out now, but I wonder if these could be "=" locations, with return statements?

Reply:
I think we want these to be internal redirects, rather than responding to the client with 301/302l if so, then I think `rewrite` is the only way to go.

Even if we're okay responding with 301/302, we want to preserve the query string and add the `service=` parameter, so using `rewrite ... redirect` is arguably nicer than `return [301]` then, since it effectively implements Apache's `RewriteRule ... [QSA]` behaviour. Else, we'd need something like:

```
location = /.well-known/host-meta.json {
    return /public.php?service=host-meta-json&$args;
}
```

This has the property that if `$args` is empty, there is a trailing `&` in the redirect — I don't know if this conforms to RFCs regarding URLs/URIs; if it does not, we cannot necessarily guarantee that clients will handle this correctly (though I expect that this isn't an issue in practice).

Reply:
Thanks for the thoroughly explanation! I agree internal redirect is best.